### ISSUE TITLE
Chers is much slower than Rustic

### ISSUE THREAD

**Issue by mvanthoor** [giusgad/chers#2]
Hi,

Rustic author here. I accidentally found your engine when searching for Rust-based chess engines and/or references to the Rustic website or GitHub repository. I can recognize many parts in your code base that are similar to Rustic. Thanks for extending the credits :)

As I always do when I find an engine that references mine, I've run a test against my current beta-version of Rustic 4. I noticed that in the middle game, Rustic is 3-5x faster than Chers depending on the position. Every speed doubling (either by being faster or by running on faster hardware) will yield about 75 Elo. So a speed difference of 4 times will make Rustic 150 Elo stronger than Chers all other things being equal. I don't know if the engine is still in development (latest commit was about a year ago), but this may be worth looking into. You're leaving at least 150 Elo on the table.

However, there's a much bigger problem with regarding to time control. I submitted a different issue for that.

**Comment by giusgad** [comment]
Hi! I built this engine purely because I wanted to try and learn something interesting as a toy project, so I moved on to other things after some time and I'm not currently working on it.
Your advice is still very much appreciated, thank you for taking the time to test things out and letting me know, if I end up picking the project back up I will definitely focus on your suggestions.

Also thank you for your great work on rustic, I think both the engine and the wiki are perfect learning resources :)

**Comment by tissatussa** [comment]
> ..Rustic is 3-5x faster than Chers depending on the position..

i guess you think this is due to the way the engine is coded, but doesn't speed also depend on the pruning method ? I mean when many nodes are skipped, for they are considered bad continuations, the engine will reach higher depth faster ..

i did a quick test with this position, which has a surprising bestmove :
r2q1rk1/ppp2ppp/2n5/3Nn3/1b6/3B1PB1/PP3P1P/R2Q1RK1 b - - 0 13

look at the number of calculated nodes :

![Bd2-chers](https://github.com/user-attachments/assets/a95db77f-25c4-46ca-9d33-82ade24492f6)

![Bd2-Rustic](https://github.com/user-attachments/assets/05662d8d-ca65-45e0-b000-d9a184e1706b)

SPOILER : solution is Bd2 !

NOTE: Chers v0.8.0 does not function in SCID, so i tested with CuteChess.
The eval is negated regarding CuteChess / SCID -- Black is better here.

Both engines find the solution at depth 10, but indeed Rustic is about 5x faster.

[ i'm on Xubuntu 22.04 ]

**Comment by mvanthoor** [comment]
@tissatussa:

Yes, pruning techniques would make a difference. It looks like Chers uses a basic alpha-beta function + transposition table at this point; the only difference is that Rustic includes Principal Variation Search. That would make a difference, but certainly not a 5x speed difference. PVS will gain 10-30 Elo depending on how good the move ordering is. (Chers has some extra parameters for move ordering, but they'll probably not have a huge impact compared to the TT-move ordering.)

I've run my own test so I can compare results on the same machine. Here are the results from each run. First is Chers:

```
marcel@workstation23:~/Code/chers/target/release$ ./chers 
uci
id name Chers
id author Giuseppe Gadola
option name Hash type spin default 128 min 1 max 32768
option name EarlyStop type check default true
option name DbgUnicode type check default true
option name Overhead type spin default 200
uciok
ucinewgame
info string Invalid command
position fen r2q1rk1/ppp2ppp/2n5/3Nn3/1b6/3B1PB1/PP3P1P/R2Q1RK1 b - - 0 13
go infinite
info depth 1 seldepth 9 score cp 469 nodes 103 nps 0 hashfull 0 time 0 pv d8d5 d3h7 
info depth 2 seldepth 11 score cp 85 nodes 3007 nps 0 hashfull 0 time 0 pv d8d6 d3f5 
info depth 3 seldepth 14 score cp 106 nodes 8677 nps 4338500 hashfull 0 time 2 pv d8d6 g3e5 c6e5 
info depth 4 seldepth 15 score cp 91 nodes 29433 nps 4204714 hashfull 0 time 7 pv d8d6 d3e4 a8d8 d1c2 
info depth 5 seldepth 20 score cp 88 nodes 232079 nps 4219618 hashfull 2 time 55 pv d8d6 d3e4 a8e8 d1b3 a7a5 g3e5 e8e5 
info depth 6 seldepth 20 score cp 66 nodes 779022 nps 3818735 hashfull 11 time 204 pv b4c5 d3e4 f7f5 g3e5 c6e5 f3f4 f5e4 f4e5 
info depth 7 seldepth 24 score cp 76 nodes 4938836 nps 3925943 hashfull 70 time 1258 pv b4d6 f3f4 e5d3 d1d3 d8d7 d3b3 d7f5 
info depth 8 seldepth 26 score cp 83 nodes 16194739 nps 3725498 hashfull 262 time 4347 pv b4c5 d3e4 f7f5 g3e5 f5e4 f3f4 c5d4 d1b3 c6e5 f4e5 d4e5 d5c7 
info depth 9 seldepth 30 score cp 109 nodes 69839121 nps 3674003 hashfull 833 time 19009 pv b4c5 a1c1 c5d4 d3e4 f7f5 g3e5 d4e5 f3f4 e5f4 
info depth 10 seldepth 30 score cp 155 nodes 397987509 nps 3658982 hashfull 1000 time 108770 pv b4d2 d3e4 f7f5 e4c2 f5f4 c2e4 c6d4 g3f4 f8f4 d1d2 e5f3 e4f3 d4f3
```

Second is Rustic Alpha 3:

```
marcel@workstation23:~/Code/rustic/target/release$ ./rustic -m 128 -f "r2q1rk1/ppp2ppp/2n5/3Nn3/1b6/3B1PB1/PP3P1P/R2Q1RK1 b - - 0 13" -q

d888888b                      dP   oo        
88     88                     88             
88oooo88  88    88  d8888b  d8888P dP d88888b
88    88  88    88  8ooooo    88   88 88     
88     88 88    88       88   88   88 88     
88     88  88888P  888888P    dP   dP 888888P
ooooooooooooooooooooooooooooooooooooooooooooo

Engine:    rustic 3.99.0
Author:    Marcel Vanthoor
EMail:     mail@marcelvanthoor.nl
Website:   https://rustic-chess.org/
Type:      64-bit
TT size:   128 MB
Protocol:  uci
Threads:   1

go
info depth 1 seldepth 10 score cp 375 time 0 nodes 111 nps 0 pv d8d5 d3h7
info depth 2 seldepth 10 score cp 100 time 0 nodes 1075 nps 0 pv e5d3 d1d3
info depth 3 seldepth 15 score cp 100 time 0 nodes 5159 nps 0 pv e5d3 d1d3 b4d6
info depth 4 seldepth 15 score cp 95 time 2 nodes 38985 nps 19492500 pv b4d6 d3f5 e5c4 d1b3
info depth 5 seldepth 18 score cp 100 time 9 nodes 126299 nps 14033222 hashfull 1 pv b4d6 d3f5 d8g5 f5e4 a8d8
info depth 6 seldepth 20 score cp 85 time 48 nodes 629952 nps 13124000 hashfull 6 pv b4d6 f3f4 e5d3 d1d3 f7f5 d3c4
info depth 7 seldepth 24 score cp 93 time 245 nodes 3197607 nps 13051457 hashfull 37 pv b4c5 d3h7 g8h7 d1c2 f7f5 c2c5 b7b6 c5b5 e5f3 g1g2
info depth 8 seldepth 24 score cp 95 time 855 nodes 10837403 nps 12675325 hashfull 135 pv b4c5 a1c1 c5d4 d3e4 a8b8 g3e5 d4e5 d1b3
info depth 9 seldepth 26 score cp 95 time 3314 nodes 42435241 nps 12804840 hashfull 463 pv b4c5 d3e4 f7f5 g3e5 c6e5 d1b3 f8f7 f3f4 f5e4 f4e5
info depth 10 seldepth 28 score cp 145 time 15175 nodes 184061319 nps 12129247 hashfull 975 pv b4d2 g3e5 c6e5 d3e4 f7f5 e4c2 c7c6 d5e3 d8g5 e3g2 a8d8
quit
```

The important metrics are:
- NPS: Nodes per second - brute force engine speed. Assuming nodes are counted correctly (Chers does it the same way as Rustic: directly after the QSearch call in alpha-beta to prevent nodes being counted double, and directly on entering the QSearch function). As both engines count NPS the same, they can be used to compare brute force speed.
- nodes: This number is the number of nodes searched while coming up with the the variation. This number gets lower the better the pruning becomes.
- time: Time to Depth. This is the actual time needed to reach this depth.

So if we compare:
- NPS: Rustic ~13 million n/s, Chers ~3.6 million n/s. This means Rustic's code is roughly 3.6x faster than Chers's. This has nothing to do with pruning.
- nodes at depth 10: Rustic 184.061.319 (~184 million) vs Chers 397.987.509 (~398 million). Chers searches 2.16x more nodes than Rustic does. Because both engines do not do anything different except for PVS (which does not have THAT big of an impact), this implies that there is something wrong with either the transposition table, or move ordering; the engine either doesn't put the moves correctly into the TT, or it cannot find them again when they're in there, or there's a bug in the move ordering function which doesn't select select the TT move to be tried first if there is any.
- time: Rustic 15175 (15.2 seconds) vs Chers 108770 (108.8 seconds). Rustic is roughly [b]7.2x[/b] faster than Chers.

The reason is obviously because Rustic searches 2.16x LESS nodes and does it 3.5x as fast. (at least in this position.) Double the speed for a chess engine means it gains about 70 Elo. At 8x the speed difference (2x2x2) this would mean a 210 Elo difference between Rustic and Chers. I've run a one-on-one match, and the 7x speed difference is shown in the roughly 185 Elo advantage Rustic has.

It would be best to first look into the difference between the node counts (fix the TT/move ordering), and then run a profiler such as Perf+HotSpot to see where the engine can be made faster. If the engine would be developed further as it is now, it will either stay at a 185 Elo disadvantage against comparable engines (at best), or newly developed functions will not yield the amount of playing strength they'd normally do in other engines (at worst).

**Comment by mvanthoor** [comment]
>Hi! I built this engine purely because I wanted to try and learn something interesting as a toy project, so I moved on to other things after some time and I'm not currently working on it.

Fair enough. I wrote Rustic and the very first draft of the book during the pandemic, but after that, life got in the way. I only worked on the engine an hour here, an hour there. However, this Christmas I've made good progress in my refactoring for version 4.0, and several chapters and sections have been added to the book.

>Your advice is still very much appreciated, thank you for taking the time to test things out and letting me know, if I end up picking the project back up I will definitely focus on your suggestions.

Good luck. I've written a fairly extensive post above. Maybe I'll look into some of these issues if I have the time. With some refactoring and probably bug-fixing, Chers could make up a large part of the 185 Elo difference to Rustic Alpha 3.

>Also thank you for your great work on rustic, I think both the engine and the wiki are perfect learning resources :)

Glad it was of use to you. I'm doing this purely for me, myself and I, because I intend (at some point) to port Rustic into the Ada programming language (as a second engine) and I don't want to have to research everything again. (Especially magic bitboards where huge pain in the neck. Most people start off with a fairly good explanation and then devolve into gibberish. I hope I can do this better, going by my own code and comments...)

Also, Rustic has now, for version 4, been refactored into librustic+engine, so I can use the chess playing part as a backend for other chess-related software. I need some documentation for that too, so I don't forget half of this stuff by the time I pick it up again.
